src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c

```diff
@@ -10,9 +10,219 @@
 #include "core/ucc_service_coll.h"
 #include "tl_mlx5.h"
+#include "tl_mlx5_mcast_helper.h"
+#include "tl_mlx5_mcast_rcache.h"
 
-ucc_status_t ucc_tl_mlx5_mcast_context_init(ucc_tl_mlx5_mcast_context_t    *context, /* NOLINT */
-                                            ucc_tl_mlx5_mcast_ctx_params_t *mcast_ctx_conf /* NOLINT */)
+#define UCC_TL_MLX5_MCAST_MAX_MTU_COUNT 5
+int mtu_lookup[UCC_TL_MLX5_MCAST_MAX_MTU_COUNT][2] =
+                        {{256, IBV_MTU_256}, {512, IBV_MTU_512},
+                        {1024, IBV_MTU_1024}, {2048, IBV_MTU_2048},
+                        {4096, IBV_MTU_4096}};
+
+ucc_status_t ucc_tl_mlx5_mcast_context_init(ucc_tl_mlx5_mcast_context_t    *context,
+                                            ucc_tl_mlx5_mcast_ctx_params_t *mcast_ctx_conf)
 {
+    ucc_status_t            status        = UCC_OK;
+    struct ibv_device     **device_list   = NULL;
+    struct ibv_device      *dev           = NULL;
+    char                   *devname       = NULL;
+    int                     is_ipv4       = 0;
+    struct sockaddr_in     *in_src_addr   = NULL;
+    struct rdma_cm_event   *revent        = NULL;
+    char                   *ib            = NULL;
+    char                   *ib_name       = NULL;
+    char                   *port          = NULL;
+    int                     active_mtu    = 4096;
+    int                     max_mtu       = 4096;
+    ucc_tl_mlx5_mcast_coll_context_t *ctx = NULL;
+
+    struct ibv_port_attr    port_attr;
+    struct ibv_device_attr  device_attr;
+    struct sockaddr_storage ip_oib_addr;
+    struct sockaddr_storage dst_addr;
+    int                     num_devices;
+    char                    addrstr[128];
+    ucc_tl_mlx5_context_t  *mlx5_ctx;
+    ucc_base_lib_t         *lib;
+    int                     i;
+    int                     user_provided_ib;
+
+    ctx = &(context->mcast_context);
+    memset(ctx, 0, sizeof(ucc_tl_mlx5_mcast_coll_context_t));
+    memcpy(&ctx->params, mcast_ctx_conf, sizeof(ucc_tl_mlx5_mcast_ctx_params_t));
+
+    mlx5_ctx = ucc_container_of(context, ucc_tl_mlx5_context_t, mcast);
+    lib      = mlx5_ctx->super.super.lib;
+    ctx->lib = lib;
+
+    device_list = ibv_get_device_list(&num_devices);
+    if (!device_list) {
+        tl_debug(lib, "no ib devices available");
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto error;
+    }
+
+    if (num_devices >= 1 && !strcmp(mcast_ctx_conf->ib_dev_name, "")) {
+        dev          = device_list[0];
+        devname      = (char *)ibv_get_device_name(dev);
+        ctx->devname = ucc_malloc(strlen(devname)+16, "devname");
+        if (!ctx->devname) {
+            status = UCC_ERR_NO_MEMORY;
+            goto error;
+        }
+        strncpy(ctx->devname, devname, strlen(devname));
+        strncat(ctx->devname, ":1", 16);
+        user_provided_ib = 0;
+    } else {
+        /* user has provided the devname now make sure it is valid */
+        for (i = 0; device_list[i]; ++i) {
```

Comment:
- because we are not using mlx5dv.h in our mcast code.
- same as above - also it gives flexibility to us to choose different devices for multicast and alltoall 
- I removed that check and added``` ib_valid ``` variable.